### ISSUE TITLE
Question about the unsupervised stage training process.

### ISSUE THREAD

**Issue by FunkyKoki** [browatbn2/3FabRec#7]
Thanks for your stimulating work!

I am curious about the unsupervised training process, so I read the related source code. But there is something I cannot understand and I will describe as below:
As I can see, the parameters updating follow the procedure: L_enc for updating D_z (4 iters once) and Encoder (2 iters once); L_adv for updating D_x (2 iters once); L_rec+L_cs+L_adv for updating Encoder & Generator (1 iter once). I hope this is right. 😄 
Here comes the question.
Firstly, L_enc for updating Encoder leaves gradient information coming form L_enc which is not useful afterwards but I cannot find any code zero_grad the gradient in Encoder. (The leaving gradient may even damage the training after ❔ ) Since you have done
`self.saae.zero_grad()`
in train_aae_unsupervised.py in line 228, I think that `self.saae.P.zero_grad()` in line 262 can be changed to `self.saae.Q.zero_grad()`. Then I think it will make sense.
Secondly, I would like to ask why the updating frequency of Discriminator is set to be slower than that of Generator?

Hope to hear from you soon~ 👍

**Comment by browatbn2** [comment]
Hi, thanks for your interest and diving into the code :) 

I think I know what you mean. We can't erase the gradients after L_enc yet because these gradients are also required for the other losses. If we call `self.saae.Q.zero_grad()` in line 262, L_rec etc would only affect the decoder. In that case we would have to send images through the encoder again. This approach is possible and I tried it but it was less stable and more computationally expensive.
Actually looking at `self.saae.P.zero_grad()`, I think it's redundant and can be removed.

The discriminators are updated slower than the generators since they are smaller and tend to train faster. This worked well in practice and saves a bit of computation time.

**Comment by FunkyKoki** [comment]
About the second question, I understand and thanks.
 However, for the first question, I am still confused about your answer. I can understand your purpose. In train_aae_unsupervised.py line 256 to 257, by setting self.args.train_encoder as True ( and this is default setting), z_sample still carries the related computation graph route. So the related loss L_rec, L_cs and L_adv related to z_sample can successfully backwards to the Encoder network. 
What I want to express is the `self.saae.Q.zero_grad()` is necessary because the `update_encoding` function has used the `loss_E` to backward and update the parameters in Encoder.
Hope you can understand.

**Comment by browatbn2** [comment]
Hm, I'm not sure if I understand the problem correctly. From what I can see is that we can't call zero_grad() at this point. 

Did you see the second backward in line 384? This one wouldn't have gradients to update if we call zero_grad on the encoder. To keep the gradients from the first backward in `update_encoding`, we call it with `retain_graph=True`. This keeps the gradients intact. Alternatively, you could also just put everything into one loss term and run only one backward at the end.
 
Yes, z_sample is part of the computation graph but it's only one part of it. Encoder weights and gradients are not stored in z_sample.

**Comment by FunkyKoki** [comment]
Did you mean line 304 in train_aae_unsupervised.py?
I doubt my understanding, but from my perspective, using `retain_graph=True` is to keep the computation graph, not the relevant gradient. Since loss_E from line 285 to 287 has backward and produce the corresponding gradient, and `self.optimizer_E` has `step` utilizing these produced gradient, these gradient should not be kept anymore, that's why I say the `self.saae.Q.zero_grad()` is necessary.
I want to claim that `retain_graph=True` is not used to keep the gradient, it is used to keep the computation graph.

**Comment by browatbn2** [comment]
Oh, yes, sorry, I meant line 304.

Ah, ok, now I got you. I think you might be right. The issue is basically that we call `self.optimizer_E` twice. So either remove the first optimizer update or erase the gradients after the first backward?

**Comment by FunkyKoki** [comment]
Yeah 😸 
These two methods would all be OK. If we choose to remove the first optimizer update (i.e. `self.optimizer_E.step()` in line 287 in aae_training.py), then `update_encoding` need to output the calculated loss and add to the final loss. But the second way may be more easy to complete.

By the way, this may not influence the network training a lot. I just want to justify the details. 🤪 hhhh

**Comment by browatbn2** [comment]
Cool, thanks a lot for your efforts and spotting this! :thumbsup:

**Comment by FunkyKoki** [comment]
Ah, I forgot an important question 🤣 
Why did you do `err_fake = err_fake[sklearn.utils.shuffle(range(len(err_fake)))]` in line 317 of aae_training.py? Is is helpful to improve the performance? Is there any explanation to do this?
Thanks~

**Comment by browatbn2** [comment]
Ah, I forgot to take that out. They initial idea was to shuffle the fakes so they are not aligned with the reals. As this doesn't affect the outcome of loss_D in line 319, however, it doesn't make any sense. :sweat_smile:

**Comment by FunkyKoki** [comment]
Alright, thanks anyway~ You are so enthusiastic!

**Comment by FunkyKoki** [comment]
Hello, I saw you have updated the related code with commit 'change P.zero_grad() to E.zero_grad() in decoding phase' . However, as I can see, the encoder in the AAE is called `Q`, so maybe it should be `self.saae.Q.zero_grad()`.

**Comment by browatbn2** [comment]
Ah, darn, yes of course 😆
